frontend/src/components/widgets/DataFrame/hooks/useColumnLoader.ts

```diff
@@ -0,0 +1,229 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { merge } from "lodash"
+
+import { Quiver } from "src/lib/Quiver"
+import { Arrow as ArrowProto } from "src/autogen/proto"
+import { notNullOrUndefined, isNullOrUndefined } from "src/lib/utils"
+import { logWarning, logError } from "src/lib/log"
+
+import {
+  getColumnTypeFromArrow,
+  getAllColumnsFromArrow,
+  getEmptyIndexColumn,
+} from "src/components/widgets/DataFrame/arrowUtils"
+import {
+  BaseColumn,
+  BaseColumnProps,
+  ObjectColumn,
+  ColumnTypes,
+  ColumnCreator,
+} from "src/components/widgets/DataFrame/columns"
+
+// Using this ID for column config will apply the config to all index columns
+export const INDEX_IDENTIFIER = "index"
+// Prefix used in the config column mapping when referring to a column via the numeric position
+export const COLUMN_POSITION_PREFIX = "col:"
+
+/**
+ * Options to configure columns.
+ */
+export interface ColumnConfigProps {
+  width?: number
+  title?: string
+  type?: string
+  hidden?: boolean
+  editable?: boolean
+  metadata?: Record<string, unknown>
+  alignment?: string
+}
+
+/**
+ * Apply the user-defined column configuration if supplied.
+ *
+ * @param columnProps - The column properties to apply the config to.
+ * @param columnConfigMapping - The user-defined column configuration mapping.
+ *
+ * @return the column properties with the config applied.
+ */
+export function applyColumnConfig(
+  columnProps: BaseColumnProps,
+  columnConfigMapping: Map<string | number, ColumnConfigProps>
+): BaseColumnProps {
+  if (!columnConfigMapping) {
+    // No column config configured
+    return columnProps
+  }
+
+  let columnConfig
+  if (columnConfigMapping.has(columnProps.title)) {
+    // Config is configured based on the column title
+    columnConfig = columnConfigMapping.get(columnProps.title)
+  } else if (
+    columnConfigMapping.has(
+      `${COLUMN_POSITION_PREFIX}${columnProps.indexNumber}`
+    )
+  ) {
+    // Config is configured based on the column position, e.g. col:0 -> first column
+    columnConfig = columnConfigMapping.get(
+      `${COLUMN_POSITION_PREFIX}${columnProps.indexNumber}`
+    )
+  } else if (
+    columnProps.isIndex &&
+    columnConfigMapping.has(INDEX_IDENTIFIER)
+  ) {
+    // Config is configured for the index column (or all index columns for multi-index)
+    columnConfig = columnConfigMapping.get(INDEX_IDENTIFIER)
+  }
+
+  if (!columnConfig) {
+    // No column config found for this column
+    return columnProps
+  }
+
+  // This will update all column props with the user-defined config for all
+  // configuration option that are not undefined:
+  return merge(
+    { ...columnProps },
```

Comment:
There a reason why to do this? Can't we just do
`merge(columnProps, {...}`

Reply:
We need a copy here since `merge` changes data in place. `merge(columnProps, {...}` would lead to unexpected problems.